### ISSUE TITLE
Android <9 crash - no non-static method UnityPlayerSingleton;.hidePreservedContent()

### ISSUE THREAD

**Issue by timbotimbo** [learntoflutter/flutter_embed_unity#15]
## Description
I'm reposting [this issue](https://github.com/juicycleff/flutter-unity-view-widget/issues/926) because it affects this plugin too.

On some Unity versions, android devices on Android 8 or lower will immediately crash when Unity is started.  

```
java.lang.NoSuchMethodError: no non-static method "Lcom/learntoflutter/flutter_embed_unity_android/unity/UnityPlayerSingleton;.hidePreservedContent()V"
```

## Breaking change
I tested multiple Unity versions to find the exact version that breaks this.

The crash starts at Unity **2021.3.31** and **2022.3.10**.  
Both these versions have this entry in the changelog, that seems like the culprit.
>- Android: Show the last rendered frame when the application come back from the pause state. ([UUM-30881](https://issuetracker.unity3d.com/issues/android-a-black-screen-appears-for-a-few-seconds-when-returning-to-the-game-from-the-lock-screen-after-idle-time))

As of **2022.3.19** this crash no longer happens, but there is nothing mentioned in the Unity changelog.  

It's your choice to either add a workaround that might become redundant or tell people to upgrade to 2022.3.19+

## Reproduction
1. Build and run any project using this plugin, like the example.  
2. Wait until Unity launches in Flutter app.
3. On Android <= 8 and the specified Unity versions, it will instantly crash.

## Workaround
For affected versions, you can avoid the crash by defining the missing function and calling the private function using reflection.  
This fixes the crash on Android <= 8 and doesn't change any behavior on higher android versions.  

I haven't benchmarked this, but it might introduce a some milliseconds delay on Android 9+ as you add a reflection call that isn't otherwise needed.

```kotlin
import java.lang.reflect.Method


     fun hidePreservedContent() {

        // manually call the private function using reflection.
        UnityPlayer::class.java.declaredMethods
        .find { it.name == "hidePreservedContent" }
        ?.let {
            it.isAccessible = true
            it.invoke(this)
        }
    
     }
```

## Versions
❌   indicates a crash.  

**Unity:**
2022.3.0 ✔  
2022.3.9 ✔  
2022.3.10 ❌  
2022.3.14 ❌
2022.3.15 ❌
2022.3.18 ❌
2022.3.19 ✔
2022.3.20 ✔

**Android:**
Android 6 ❌  
Android 7.1.2 ❌  
Android 8  ❌  (reported)
Android 9 ✔
Android 13 ✔

**Devices:**
This seems to happen on any device on Android <= 8.  
 - Samsung Galaxy J5, Android 6
 - Wileyfox Swift, Android 7
 - (reported) Huawei Mate 10 Lite, Android 8
 - (reported) Samsung Galaxy S7

## Logs

<details>
<summary>Log Unity 2022.3.18, Android 7</summary>

```
E/AndroidRuntime( 5807): FATAL EXCEPTION: UnityMain
E/AndroidRuntime( 5807): Process: com.learntoflutter.flutter_embed_unity_example, PID: 5807
E/AndroidRuntime( 5807): java.lang.Error: FATAL EXCEPTION [UnityMain]
E/AndroidRuntime( 5807): Unity version     : 2022.3.18f1
E/AndroidRuntime( 5807): Device model      : Wileyfox Wileyfox Swift
E/AndroidRuntime( 5807): Device fingerprint: Wileyfox/Swift/crackling:7.1.2/N2G48B/eeea277114:user/release-keys
E/AndroidRuntime( 5807): CPU supported ABI : [arm64-v8a, armeabi-v7a, armeabi]
E/AndroidRuntime( 5807): Build Type        : Release
E/AndroidRuntime( 5807): Scripting Backend : IL2CPP
E/AndroidRuntime( 5807): Libs loaded from  : lib/arm64
E/AndroidRuntime( 5807): Strip Engine Code : true
E/AndroidRuntime( 5807): 
E/AndroidRuntime( 5807): Caused by: java.lang.NoSuchMethodError: no non-static method "Lcom/learntoflutter/flutter_embed_unity_android/unity/UnityPlayerSingleton;.hidePreservedContent()V"
E/AndroidRuntime( 5807): 	at com.unity3d.player.UnityPlayer.nativeRender(Native Method)
E/AndroidRuntime( 5807): 	at com.unity3d.player.UnityPlayer.-$$Nest$mnativeRender(Unknown Source)
E/AndroidRuntime( 5807): 	at com.unity3d.player.UnityPlayer$F$a.handleMessage(Unknown Source)
E/AndroidRuntime( 5807): 	at android.os.Handler.dispatchMessage(Handler.java:98)
E/AndroidRuntime( 5807): 	at android.os.Looper.loop(Looper.java:154)
E/AndroidRuntime( 5807): 	at com.unity3d.player.UnityPlayer$F.run(Unknown Source)
```
</details>

**Comment by jamesncl** [comment]
Thanks for your usual high quality bug report!

I've just pushed your workaround to a [new branch bug/issue15](https://github.com/learntoflutter/flutter_embed_unity/tree/bug/issue15) with a conditional check so it only runs the reflection if it's <Android 9. I don't have any Android 8 devices to test this with, would appreciate if you could check this fixes the issue (I've also updated the dependencies to point to local paths, so you should just be able to run the example project in the android sub-package)

**Comment by timbotimbo** [comment]
The issue15 branch fixes the crash on my Android 7 device, but actually breaks on Android >= 9.

https://github.com/learntoflutter/flutter_embed_unity/assets/11444698/41b28045-6086-473f-8c9c-8216905203c4

##
- Android <9:
  Unity can't access the hidePreservedContent function and crashes.  
  - Adding the function definition fixes the crash.  
  - Using reflection to call hidePreservedContent fixes any resume issues.  
  
- Android > 8:
  Unity does manage to call the internal hidePreservedContent function.  
  - Adding the function definition hides the original and turns this into a no-op and the paused frame placeholder will never get removed.  (see video above)   
  - Adding the reflection call makes this work too, but you're doing a redundant reflection call. 
  
Ideally you would want to do conditional compilation like possible in C#, but is seems like kotlin doesn't support this.
```
#if android < 9
 fun hidePreservedContent() { ... }
#endif
```

##

I've also seen someone on the flutter_unity_widget discord having freezing issues on resume (like in my video), that were resolved once he upgraded from 2022.3.16 to 2022.3.20.  
This sounds like it is caused by the same issue except that his app didn't crash.  

He got the issue on an Android 8 phone and Android 11 emulator.

So there might be more devices affected in a different way, but I haven't been able to reproduce that yet.

**Comment by jamesncl** [comment]
Thanks for the update - I found an old Motorola with Android 5.1 and I can now reproduce the issue, and I see what you mean about it breaking on Android > 8.

As far as I know conditional compilation based on Android SDK isn't possible because at compile time you don't know the version of Android, it has to be a runtime check.

So it looks like the options are:
1. always use reflection regardless of SDK version. I agree it's not ideal, but also not the end of the world - I imagine this function isn't called a lot. It might be an acceptable tradeoff for stability for users
2. figure out what this function actually does and replicate it

For 2, I took a copy of **/Applications/Unity/Hub/Editor/2022.3.20f1/PlaybackEngines/AndroidPlayer/Variations/mono/Development/Classes/classes.jar**, unzipped, and opened in Android Studio (it decompiles class files). The decompiled version of the function in `UnityPlayer.class` is very simple, it is just this:

```java
    private void hidePreservedContent() {
        c var1;
        var1 = new c.<init>(this);
        this.runOnUiThread(var1);
    }
```

I checked Unity 2022.3.9, this function does not exist. It does exist in 2022.3.13 and 2022.3.20, and is the same in both.

The only problem is I can't figure out what `c` might refer to. Something that you create, pass UnityPlayer instance to it, and run it. If anyone can figure out what this is, we can probably just replicate the function and avoid using reflection. On the other hand, it may make the package more brittle to changes in future versions of Unity (i.e. if this function changes), so maybe refection is the better choice anyway?

**Comment by timbotimbo** [comment]
I feel like replicating the internals is just looking for trouble. I wouldn't be surprised if it breaks again without even a changelog mention in the next Unity release.

A third option would be to make people use >= 2022.3.19 as it seems to be fixed there. But given that there is no bugfix documented anywhere, this might just be an accidental fix that breaks again in the future.

**Comment by jamesncl** [comment]
On balance I think it's best to go with your third option. I've just published version 1.0.4 with these details in the README. Thanks for your help!
